### PR TITLE
Try sending small key frames to clear decoded buffer

### PR DESCRIPTION
**Pull request by boks1971**
Problem:
--------
With transceiver re-use, client disables/enables tracks.
With video, this retains the last picture and when a new track
starts, there is a brief moment when the old stream is displayed
till there is data from new stream to decode and display.

Fix:
---
Send small key frames before closing DownTrack to try and clear
the decoder display buffer.

Testing:
--------
Tried with Chrome/Safari/Firefox and they worked. But, very rarely,
the last frames do not seem to show up. In fact, 6 frames are sent
and webrtc internals (in Firefox) reports anywhere from 1 - 6 frames
at the small resolution. Unclear as to why it does not get all the
frames or why it reports less than 6. A not so small percentage of
times (maybe 1 in 15 - 20), have seen no small frame reported at all.

TODO:
----
- Have to support more video codecs
- Would this be an issue for audio also? Should we send something to handle that?
  Probably not necessary as video is more jarring.

### CODE REVIEW

File: pkg/sfu/downtrack.go [hunk 1]
```diff
@@ -131,7 +132,12 @@ func NewDownTrack(c webrtc.RTPCodecCapability, r Receiver, bf *buffer.Factory, p
 		receiver:      r,
 		codec:         c,
 		munger:        NewMunger(),
-	}, nil
+	}
+	if strings.ToLower(c.MimeType) == "video/vp8" {
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:136]: Creating a munger here every time. This should take care of the multi-threading issue you mentioned @davidzhao . It is a bit of overkill for simple tracks without temporal layers as this is really needed only when sending the last blank frames, but should not be a big burden. But, something to keep an eye on.

File: pkg/sfu/downtrack.go [hunk 2]
```diff
@@ -859,7 +942,7 @@ func (d *DownTrack) writeSimpleRTP(extPkt *buffer.ExtPacket) error {
 	}
 	if d.sequencer != nil {
 		meta := d.sequencer.push(extPkt.Packet.SequenceNumber, newSN, newTS, 0, extPkt.Head)
-		if meta != nil && d.vp8Munger != nil {
+		if meta != nil && translatedVP8 != nil {
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:945]: The translated header can be `nil` for padding only packets. So, checking for that.

File: pkg/sfu/downtrack.go [hunk 3]
```diff
@@ -1344,17 +1421,15 @@ func (m *Munger) UpdateAndGetSnTs(extPkt *buffer.ExtPacket) (uint16, uint32, err
...
-			for lostSN := m.highestIncomingSN + 1; lostSN <= uint16(buffer.MaxSN-1); lostSN++ {
-				m.missingSNs[lostSN] = m.snOffset
-			}
-			for lostSN := uint16(0); lostSN < extPkt.Packet.SequenceNumber; lostSN++ {
-				m.missingSNs[lostSN] = m.snOffset
-			}
+			lossEndSN = int(extPkt.Packet.SequenceNumber) - 1 + buffer.MaxSN
+		}
+		for lostSN := lossStartSN; lostSN <= lossEndSN; lostSN++ {
+			m.missingSNs[uint16(lostSN&0xffff)] = m.snOffset
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:1432]: Cleaning this up to use `int` range as `uint16` could have caused trouble in wrap around case check.

Comment (cnderrauber) [reply pkg/sfu/downtrack.go:1432]: Maybe we can add extSN field to extPacket, that can avoid check seq wrap everywhere.

**Review by cnderrauber — APPROVED** [review]
looks good

**Comment by boks1971** [comment]
> Maybe we can add extSN field to extPacket, that can avoid check seq wrap everywhere.

@cnderrauber I want to move a bunch of these to the `receiver` so that each `DownTrack` does not need to do this. Will come back to it some time.

File: pkg/sfu/downtrack.go [hunk 4]
```diff
@@ -387,6 +393,83 @@ func (d *DownTrack) WritePaddingRTP(bytesToSend int) int {
...
+		if !d.munger.IsOnFrameBoundary() {
+			frameEndNeeded = true
+		}
+
+		sn, ts, err := d.munger.UpdateAndGetPaddingSnTs(frameEndNeeded)
+		if err != nil {
+			return err
+		}
+
+		adjustedTs := ts + uint32(i+1)*(d.codec.ClockRate/30) // assume 30 fps
```
Comment (davidzhao) [inline pkg/sfu/downtrack.go:417]: I assume this would work as well even when fps is lower? i.e. larger gap?

File: pkg/sfu/downtrack.go [hunk 5]
```diff
@@ -387,6 +393,83 @@ func (d *DownTrack) WritePaddingRTP(bytesToSend int) int {
...
+		blankVP8, err := d.vp8Munger.UpdateAndGetPadding(!frameEndNeeded)
+		if err != nil {
+			return err
+		}
+
+		// 1x1 key frame
+		// Used even when closing out a previous frame. Looks like receivers
+		// do not care about content (it will probably end up being an undecodable
+		// frame, but that should be okay as there are key frames following)
+		vp8PayloadHeader := []byte{0x10, 0x02, 0x00, 0x9d, 0x01, 0x2a, 0x01, 0x00, 0x01, 0x00, 0x0b, 0xc7, 0x08, 0x85, 0x85, 0x88, 0x85, 0x84, 0x88, 0x3f, 0x82, 0x00, 0x0c, 0x0d, 0x60, 0x00, 0xfe, 0xe6, 0xb5, 0x00}
```
Comment (davidzhao) [inline pkg/sfu/downtrack.go:446]: nit:  could we move this to a const?

**Review by davidzhao — APPROVED** [review]
lg!

**Comment by davidzhao** [comment]
this seemed to have broke a bunch of publishing related tests.

File: pkg/sfu/downtrack.go (hunk 4 above)
Comment (boks1971) [reply pkg/sfu/downtrack.go:421]: Yeah, it should be fine as it is just cleaning the buffer :-)

File: pkg/sfu/downtrack.go (hunk 5 above)
Comment (boks1971) [reply pkg/sfu/downtrack.go]: Done in this commit @davidzhao - https://github.com/livekit/livekit-server/pull/179/commits/d0af4b3d1e7aa926f1051889108bf65d8535bf63

**Comment by boks1971** [comment]
@davidzhao 

> this seemed to have broke a bunch of publishing related tests.

Yeah, I broke the simple video track with this change. As the VP8 munger is used for simple tracks also now, had to set up a packet factory buffer for use in munging. 

Also made some small changes in DownTrack to make the blank frame function private - 
https://github.com/livekit/livekit-server/pull/179/commits/9fd8875b943d4ba7c9d849062473762c471aabe4

**Pull request merged**